Add date column to the built-in search engine

ducalex opened the pull request:
This PR adds a date column to the built-in search engine to show when a torrent was published/uploaded on the torrent site.

When a plugin wants to show a date, it can now add a `date` entry to its result dict. The value format is a unix timestamp (an integer representing seconds since epoch).

This commit adds the date column to both the Qt UI and the WebUI and updates the nova3 prettyPrinter.

Plugins with no date support will keep working.

If this PR is accepted, I will send a PR to the plugins repo to update as many plugins as I can! Here's the updated thepiratebay for testing the PR: https://github.com/ducalex/qbittorrent-search-plugins/blob/ducalex/add-date-field-tpb/nova3/engines/piratebay.py

Related open issues: #1463


Still to do are translations (a single additional string "Date"), am I expected to update all translations?

The guidelines suggest I include a screenshot (desktop top, webui bottom):
![Untitled](https://github.com/qbittorrent/qBittorrent/assets/13711380/66abf519-fc5b-4969-a304-367aa80d1377)

Piccirello commented:
This is a nice addition! I think a more descriptive name could be useful, like "Uploaded On" or "Created On".

I'll note that I personally prefer "At" to "On" (i.e. "Uploaded At") since this field includes the time, however the Transfers table already uses On for this use case (e.g. "Added On', "Completed On").

ducalex commented:
> This is a nice addition! I think a more descriptive name could be useful, like "Uploaded On" or "Created On".
> 
> I'll note that I personally prefer "At" to "On" (i.e. "Uploaded At") since this field includes the time, however the Transfers table already uses On for this use case (e.g. "Added On', "Completed On").

I struggled to choose a name! I've considered Added, Uploaded, Created, and Published.  I quickly dismissed Created because the torrent itself also has a creation date that is unrelated to when i twas sent to the website so it might be misleading (Though if you think Created is best, I'll still go with it of course). I've looked at what the torrent websites usually name that column and "Added" or "Added On" seemed the most common, followed by simply "Date" or "Time".


There's also the question of all the other symbols added by this PR:

- SearchResultColumn enum name (currently `PL_DATE`)
- SearchResult member variable name (currently `date`)
- SearchSortModel SearchColumn enum name (currently `DATE`)
- The key used in json serialization for the WebUI (currently `date`)
- The key used in the search plugins (currently `date`)


Do you foresee a future where more than one "date" might be returned by a search result? If so maybe I should rename all of them to something better, but what?

glassez commented:
> I struggled to choose a name! I've considered Added, Uploaded, Created, and Published

I would vote for Published.

glassez commented:
> maybe I should rename all of them to something better, but what?

Accordingly to chosen name, e.g. PL_PUBLICATION_DATE.

Piccirello commented:
My personal choice is for Uploaded On. If we do go with Published instead, I'd strongly push for appending "On" to make clear that it's a date.

> Do you foresee a future where more than one "date" might be returned by a search result? If so maybe I should rename all of them to something better, but what?

It's unlikely that there will be, but it's also notoriously difficult to predict future improvements. I never would have predicted this date field! The safest option would be to rename the relevant symbols once you decide on a new column name (e.g. `uploaded_on`).

glassez commented:
> If we do go with Published instead, I'd strongly push for appending "On" to make clear that it's a date.

It was implied that any of the suggested options would have been used with the addition of "On", wasn't it? At least that's how I thought of it. So yes, I vote for "Published On".

ducalex commented on src/gui/search/searchjobwidget.cpp:
```diff
@@ -533,6 +534,7 @@ void SearchJobWidget::appendSearchResults(const QVector<SearchResult> &results)
         setModelData(SearchSortModel::SIZE, Utils::Misc::friendlyUnit(result.fileSize), result.fileSize, (Qt::AlignRight | Qt::AlignVCenter));
         setModelData(SearchSortModel::SEEDS, QString::number(result.nbSeeders), result.nbSeeders, (Qt::AlignRight | Qt::AlignVCenter));
         setModelData(SearchSortModel::LEECHES, QString::number(result.nbLeechers), result.nbLeechers, (Qt::AlignRight | Qt::AlignVCenter));
+        setModelData(SearchSortModel::DATE, QLocale().toString(result.date.toLocalTime(), QLocale::ShortFormat), result.date);
```
It's blank.

ducalex replied:
Would you prefer it shows `Unknown`? Perhaps we should add a new helper `Utils::Misc::friendlyDate` to handle it?

glassez replied:
> Would you prefer it shows `Unknown`?

I would keep it blank for now.

xavier2k6 commented:
> > I struggled to choose a name! I've considered Added, Uploaded, Created, and Published
> 
> I would vote for Published.

Another vote in favor of Published On.

Pentaphon commented:
>Another vote in favor of Published On.

This is the only right solution for the new column.

ducalex commented:
I've made the changes for "Published On" and renamed struct/enum members to pubDate/PUB_DATE (I went with abbreviation because other sibling members also used abbreviation, like DESC. but let me know if you want a full publicationDate/PUBLICATION_DATE).

glassez commented on src/base/search/searchhandler.cpp:
```diff
@@ -176,7 +177,7 @@ bool SearchHandler::parseSearchResult(const QStringView line, SearchResult &sear
     const int nbFields = parts.size();
 
-    if (nbFields < (NB_PLUGIN_COLUMNS - 1)) return false; // -1 because desc_link is optional
+    if (nbFields < PL_ENGINE_URL + 1) return false; // Anything after ENGINE_URL is optional
```
```suggestion
    if (nbFields <= PL_ENGINE_URL) return false; // Anything after ENGINE_URL is optional
```

glassez commented on src/webui/api/searchcontroller.cpp:
```diff
@@ -301,6 +302,7 @@ int SearchController::generateSearchId() const
  *   - "nbLeechers"
  *   - "siteUrl"
  *   - "descrLink"
+ *   - "date"
```
```suggestion
 *   - "pubDate"
```

glassez commented:
> Code is identical but I know this forces you to review again, I apologize for the trouble!

Well, GitHub turned out to be smart enough to figure out this, so even the previous approval was not dismissed.

stalkerok commented:
> Published On

Very strange decision to choose this particular name, it is not used anywhere on trackers, usually it is “Added” “Uploaded”

Pentaphon commented:
> Very strange decision to choose this particular name, it is not used anywhere on trackers, usually it is “Added” “Uploaded”

The search engine is not a tracker so "published on" works for all sources regardless of what they use.

Chocobo1 commented on src/searchengine/nova3/novaprinter.py:
```diff
@@ -26,12 +26,16 @@
-                        str(dictionary["leech"]), dictionary["engine_url"]))
-    if 'desc_link' in dictionary:
-        outtext = "|".join((outtext, dictionary["desc_link"]))
+    outtext = "|".join((
```
The `VERSION:` on top of this file should be incremented.

ducalex replied:
You're right! I'm guessing `1.48` is fine? Or should it be 2.0 since it adds fields to the output?

ducalex replied:
According to the history it's always been bumped by 0.1 even when format changes slightly, so I'll do the same.

stalkerok commented:
> The search engine is not a tracker so "published on" works for all sources regardless of what they use and best of all, makes the most sense.

Nevertheless, you search by trackers and the displayed date is also taken from the tracker. IMO, I would not deviate from the variant used by trackers.

Pentaphon commented:
> Nevertheless, you search by trackers and the displayed date is also taken from the tracker. IMO, I would not deviate from the variant used by trackers.

What they use varies. It's best to pick one thing for our application that makes sense. A torrent is published on a site so "published on" makes sense. If you can come up with a column name that people think is better than "published on", I'd like to see it.

stalkerok commented:
Don't have to make anything up, most often it's “Added” with or without “On”.

stalkerok commented:
It doesn't work for me, before the latest changes it worked fine.

glassez commented:
> It doesn't work for me, before the latest changes it worked fine.

What do you mean?
I just downloaded the build artifact from GitHub and it works fine. It just adds an empty "published on" column, which is quite expected, because plugins still do not provide the information to fill it.

stalkerok commented:
> What do you mean?
> I just downloaded the build artifact from GitHub and it works fine. It just adds an empty "published on" column, which is quite expected, because plugins still do not provide the information to fill it.

I got the plugin from @ducalex https://github.com/ducalex/qbittorrent-search-plugins/blob/f53497c719d3f25fa83cf109142029808dc1a981/nova3/engines/piratebay.py

https://github.com/qbittorrent/qBittorrent/actions/runs/8756584963 Working
![2024-04-25_115318](https://github.com/qbittorrent/qBittorrent/assets/63958081/4115ce3f-264e-4ddd-85a3-7812ad22dbe1)

https://github.com/qbittorrent/qBittorrent/actions/runs/8801806693 Not working anymore
![2024-04-25_115306](https://github.com/qbittorrent/qBittorrent/assets/63958081/c69b1d6d-dd8c-4866-bc25-56f13b595350)

glassez commented:
> I got the plugin from @ducalex https://github.com/ducalex/qbittorrent-search-plugins/blob/f53497c719d3f25fa83cf109142029808dc1a981/nova3/engines/piratebay.py

He just didn't change it accordingly. It works if you change 'date' to 'pub_date'.

ducalex commented:
> He just didn't change it accordingly. It works if you change 'date' to 'pub_date'.

That's my bad, I changed the link in the description to point to the branch head.

There's a few more plugins available (you'll have to pick the appropriate file from each branch): https://github.com/ducalex/qbittorrent-search-plugins/branches though some are still WIP like EZTV.


As for the name of the column I don't have much of an opinion, I changed it to Published On (perhaps prematurely) because it seemed to go in that direction in the comments.

glassez commented:
@ducalex 
Thank you!

lunchanddinner commented:
Nice, how is this plugin going along? I think it's a very needed feature
